src/test/app/LedgerHistory_test.cpp

```diff
@@ -0,0 +1,254 @@
+                hotACCOUNT_NODE, res->info().seq);
+            res->txMap().flushDirty(
+                hotTRANSACTION_NODE, res->info().seq);
+
+        }
+        res->unshare();
+
+        // Accept ledger
+        res->setAccepted(res->info().closeTime,
+            res->info().closeTimeResolution, true /* close time correct*/,
+            env.app().config());
+        lh.insert(res, false);
+        return res;
+    }
+
+    void
+    testHandleMismatch()
+    {
+        testcase("LedgerHistory mismatch");
+        using namespace jtx;
+        using namespace std::chrono;
+
+        // No mismatch
+        {
+            bool found = false;
+            Env env{*this,
+                    envconfig(),
+                    std::make_unique<CheckMessageLogs>("MISMATCH ", found)};
+            LedgerHistory lh{beast::insight::NullCollector::New(),
+                                env.app()};
+            auto const genesis = makeLedger({}, env, lh, 0s);
+            uint256 const dummyTxHash{1};
+            lh.builtLedger(genesis, dummyTxHash, {});
+            lh.validatedLedger(genesis, dummyTxHash);
+
+            BEAST_EXPECT(!found);
+        }
+
+        // Close time mismatch
+        {
+            bool found = false;
+            Env env{*this,
+                    envconfig(),
+                    std::make_unique<CheckMessageLogs>(
+                        "MISMATCH on close time", found)};
+            LedgerHistory lh{beast::insight::NullCollector::New(), env.app()};
+            auto const genesis = makeLedger({}, env, lh, 0s);
+            auto const ledgerA = makeLedger(genesis, env, lh, 4s);
+            auto const ledgerB = makeLedger(genesis, env, lh, 40s);
+
+            uint256 const dummyTxHash{1};
+            lh.builtLedger(ledgerA, dummyTxHash, {});
+            lh.validatedLedger(ledgerB, dummyTxHash);
+
+            BEAST_EXPECT(found);
+        }
+
+        // Prior ledger mismatch
+        {
+            bool found = false;
+            Env env{*this,
+                    envconfig(),
+                    std::make_unique<CheckMessageLogs>(
+                        "MISMATCH on prior ledger", found)};
+            LedgerHistory lh{beast::insight::NullCollector::New(), env.app()};
+            auto const genesis = makeLedger({}, env, lh, 0s);
+            auto const ledgerA = makeLedger(genesis, env, lh, 4s);
+            auto const ledgerB = makeLedger(genesis, env, lh, 40s);
+            auto const ledgerAC = makeLedger(ledgerA, env, lh, 4s);
+            auto const ledgerBD = makeLedger(ledgerB, env, lh, 4s);
+
+            uint256 const dummyTxHash{1};
+            lh.builtLedger(ledgerAC, dummyTxHash, {});
+            lh.validatedLedger(ledgerBD, dummyTxHash);
+
+            BEAST_EXPECT(found);
+        }
+
+        // Simulate a bug in which consensus may agree on transactions, but
+        // somehow generate different ledgers
+        for (const bool txBug : {true, false})
```

Comment:
Nit: east const (heh, I just heard "east" const and "west" const recently and thought I'd pretend to be cool)